### PR TITLE
Enable cart item images filtering, fixes #42259

### LINKED ISSUE #42259: Cart block - Third party plugin compatibility - Need replacement for woocommerce_cart_item_thumbnail hook
## Describe the bug

In the cart and checkout blocks, there is no way to replace the product image

## To reproduce

1. Go to https://demos.mklacroix.com/custom-laptop/
2. Select some options
3. Add to cart
4. Visit the cart comparison page (https://demos.mklacroix.com/cart-block-vs-cart-shortcode/)

or 

1. Use the filter `woocommerce_cart_item_thumbnail` to change the image of the cart items
2. e.g. 
```php
add_filter( 'woocommerce_cart_item_thumbnail', function() {
    return '<img src="https://user-images.githubusercontent.com/11990914/257510384-b6e0eff2-bcef-49d5-aa76-3c25c82f08ba.png">';
} );

```

## Expected behavior

The product image in the cart should be changed. Ideally there would be a PHP hook in the store API allowing to change the image. A JS hook could also be used, but it would not be ideal in my opinion, as it would require duplicating the logic which already exists in PHP.

## Screenshots

Cart Shortcode: | Cart Block:
--|--
image | image

-   WordPress version: 6.3
-   WooCommerce version: 8.0.1
-   WooCommerce Blocks version: 10.8.2
-   Site language: EN
-   Any other plugins installed: [Product configurator for WooCommerce](https://wordpress.org/plugins/product-configurator-for-woocommerce/) and add-ons. 

## Additional context

As the developper of the plugins mentioned above, this issue is part of a series of issues brought forward in an effort to make them compatible with the cart and checkout blocks.

### PR DESCRIPTION
**Pull request by synercoder**
### Changes proposed in this Pull Request:

Adding a filter hook named `woocommerce_cart_item_images` that receives 3 parameters:

 - `$product_images` An array of image objects to be filtered.
 - `$cart_item` The cart item the images belong to.
 - `$cart_item_key` The cart item key.

This way a plugin can edit cart item images on a per item basis.

### How to test the changes in this Pull Request:

Code review is probably sufficient for this change. However, you can easiliy test this change by running through the following steps:

1. Install WooCommerce
2. Activate WooCommerce
3. Add a product
4. _Optional: Add images to the product_
5. Install and activate a theme with cart-blocks support, for example storefront
6. Add the created product to the cart. You should now see either the selected images, or a placeholder in the cart
7. Create a test plugin using the following code:
```
<?php

/**
 * @wordpress-plugin
 * Plugin Name:       ReplaceCartThumbnail
 * Description:       This plugin replaces cart thumbnails with random images
 */

// Discard the original product images array, and returns a new array with a replaced image
add_filter( 'woocommerce_cart_item_images', function ($productimages, $cart_item, $cart_item_key)
{
  $image_path = "https://picsum.photos/seed/$cart_item_key/200";
  return [
    (object)[
      'id'        => (int) 0,
      'src'       => $image_path,
      'thumbnail' => $image_path,
      'srcset'    => (string)'',
      'sizes'     => (string)'',
      'name'      => 'Random product image',
      'alt'       => 'Random product image',
    ]
  ];
}, 10, 3 );
```

8. Install the test plugin and activate
9. Go back to the cart.
10. The cart should now show a random image as the thumbnail

### Changelog entry

-   [X] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [X] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [X] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Enables replacing cart item images in cart-blocks using a filter. Complementary shortcode hook already exists: woocommerce_cart_item_thumbnail.

</details>

### CODE REVIEW

**Comment by synercoder** [comment]
First time PR to Woocommerce, and PHP is not my main language. I created the fork and edit online in GitHub, so I wasn't able to run `pnpm --filter=@woocommerce/plugin-woocommerce lint:php:changes:branch` (would not even be able to do that locally, never worked with phpm before). But the change is quite small so there shouldn't be any linting errors. (But please correct me if I'm wrong).

**Comment by Konamiman** [comment]
I'm removing myself from the list of reviewers since Proton is not involved in the development of the Store API, but of course I'm available for any question about WooCommerce core development in general.

**Review by nielslange — APPROVED** [review]
Congratulations on your first PR to WooCommerce, and thanks for working on this. I've successfully tested this PR and in comparison to the code snippet I mentioned in https://github.com/woocommerce/woocommerce/issues/42259#issuecomment-2076423860, your implementation gets called only once per product image.

As for the testing instructions, instead of creating a plugin based on your example code, it's also possible to install and activate the [Code Snippets](https://wordpress.org/plugins/code-snippets/) plugin and simply add the example code there. This might be a bit easier for some testers. That said, the outcome is the same.

I'm pre-approving this PR, but would appreciate it if @opr could take another look at it.

**Comment by synercoder** [comment]
> Congratulations on your first PR to WooCommerce, and thanks for working on this. I've successfully tested this PR and in comparison to the code snippet I mentioned in [#42259 (comment)](https://github.com/woocommerce/woocommerce/issues/42259#issuecomment-2076423860), your implementation gets called only once per product image.

Thanks!

> As for the testing instructions, instead of creating a plugin based on your example code, it's also possible to install and activate the [Code Snippets](https://wordpress.org/plugins/code-snippets/) plugin and simply add the example code there. This might be a bit easier for some testers. That said, the outcome is the same.

Nice! did not know about this plugin, but makes some (future) testing on my side also easier!

> I'm pre-approving this PR, but would appreciate it if @opr could take another look at it.

Awesome to hear, and I look forward to either further feedback or seeing the PR approved/merged. :)

File: plugins/woocommerce/src/StoreApi/Schemas/V1/CartItemSchema.php [hunk 1]
```diff
@@ -79,6 +79,30 @@ public function get_item_response( $cart_item ) {
...
+		 * Filter the cart product images
+		 *
+		 * This hook allows for changing the cart item images. This is specific to the cart endpoint.
+		 *
+		 * @param array  product_images  Array of image objects, as defined in ImageAttachmentSchema
+		 * @param array  $cart_item      Cart item array.
+		 * @param string $cart_item_key  Cart item key.
+		 */
+		return apply_filters( 'woocommerce_cart_item_images', $product_images, $cart_item, $cart_item_key );
+	}
```
Comment (nielslange) [inline plugins/woocommerce/src/StoreApi/Schemas/V1/CartItemSchema.php]: I had another chat with @opr, and we believe that we should save guard the filter. If the filter does not return a valid image, the original image should be used as a fell back. Such a safeguard could look like this:

```php
	protected function get_cart_images( \WC_Product $product, array $cart_item, string $cart_item_key ) {
		// Get the images for the product
		$product_images = $this->get_images( $product );

		/**
		 * Filter the cart product images
		 *
		 * This hook allows for changing the cart item images. This is specific to the cart endpoint.
		 *
		 * @param array  product_images  Array of image objects, as defined in ImageAttachmentSchema
		 * @param array  $cart_item      Cart item array.
		 * @param string $cart_item_key  Cart item key.
		 */
		$filtered_images = apply_filters( 'woocommerce_cart_item_images', $product_images, $cart_item, $cart_item_key );

		// Return the original images if the filtered image has no thumbnail URL.
		foreach ( $filtered_images as $key => $image ) {
			if ( empty( $image->thumbnail ) || ! filter_var( $image->thumbnail, FILTER_VALIDATE_URL ) ) {
				return $product_images;
			}
		}

		// Return the filtered images.
		return $filtered_images;
	}
```

**Review by nielslange — CHANGES_REQUESTED** [review]
Hello @synercoder 👋

@opr and I looked at the changed and I requested one change regarding save guarding the filter. You can find an example of how such a safeguard could look like. Once that change had been applied, I think the PR is in a state to be merged. 😀

File: plugins/woocommerce/src/StoreApi/Schemas/V1/CartItemSchema.php (hunk 1 above)
Comment (synercoder) [reply plugins/woocommerce/src/StoreApi/Schemas/V1/CartItemSchema.php]: Would you want the guard to first filter out images without thumbnail/src, and return the remaining guarded filtered image array, or on an single filtered image failing the guard, return the original array? Cause the provided guard sample will fall back to the original array if a single image fails the guard. Might be intended, just wanted to make sure.

For example:
``` PHP
protected function get_cart_images( \WC_Product $product, array $cart_item, string $cart_item_key ) {
	// Get the images for the product
	$product_images = $this->get_images( $product );

	/**
	 * Filter the cart product images
	 *
	 * This hook allows for changing the cart item images. This is specific to the cart endpoint.
	 *
	 * @param array  product_images  Array of image objects, as defined in ImageAttachmentSchema
	 * @param array  $cart_item      Cart item array.
	 * @param string $cart_item_key  Cart item key.
	 */
	$filtered_images = apply_filters( 'woocommerce_cart_item_images', $product_images, $cart_item, $cart_item_key );

	// Return the original images if the filtered image has no thumbnail URL.
	$guarded_images = array();
	foreach ( $filtered_images as $key => $image ) {
		// Check if thumbnail is a valid url
		if ( empty( $image->thumbnail ) || ! filter_var( $image->thumbnail, FILTER_VALIDATE_URL ) ) {
			continue;
		}
		// Check if src property is a valid url
		if ( empty( $image->src) || ! filter_var( $image->src, FILTER_VALIDATE_URL ) ) {
			continue;
		}

		// Image is valid, add to resulting array
		$guarded_images[] = $image;
	}

	// If there are no valid images remaining, return original array
	if ( count($guarded_images) == 0 ) {
		return $product_images;
	}

	// Return the filtered guarded images.
	return $guarded_images;
}
```

Comment (synercoder) [reply plugins/woocommerce/src/StoreApi/Schemas/V1/CartItemSchema.php]: Also, when failing a guard check, maybe add a log for it? So plugin developers can see in logs why it is failing?

``` PHP
$logger = wc_get_logger();
foreach ( $filtered_images as $key => $image ) {
	// Check if thumbnail is a valid url
	if ( empty( $image->thumbnail ) || ! filter_var( $image->thumbnail, FILTER_VALIDATE_URL ) ) {
		$logger->warning( "After passing through woocommerce_cart_item_images filter, image with id $image->id 
 did not have a valid thumbnail property." );
		continue;
	}
	// Check if src property is a valid url
	if ( empty( $image->src) || ! filter_var( $image->src, FILTER_VALIDATE_URL ) ) {
		$logger->warning( "After passing through woocommerce_cart_item_images filter, image with id $image->id 
 did not have a valid src property." );
		continue;
	}

	// Image is valid, add to resulting array
	$guarded_images[] = $image;
}
```

Comment (nielslange) [reply plugins/woocommerce/src/StoreApi/Schemas/V1/CartItemSchema.php]: > Would you want the guard to first filter out images without thumbnail/src, and return the remaining guarded filtered image array, or on an single filtered image failing the guard, return the original array? Cause the provided guard sample will fall back to the original array if a single image fails the guard. Might be intended, just wanted to make sure.

Great catch, @synercoder. We should display images from the filter as long as they're valid. Logging invalid images, also makes sense, IMO. Could you implement these changes, and apply the coding styles?

**Comment by opr** [comment]
I'm removing myself as a reviewer until the changes @nielslange asked for have been implemented. I don't think it makes sense for me to review until then. Please ping/re-request from me when done! thanks.

File: plugins/woocommerce/src/StoreApi/Schemas/V1/CartItemSchema.php (hunk 1 above)
Comment (synercoder) [reply plugins/woocommerce/src/StoreApi/Schemas/V1/CartItemSchema.php]: @nielslange @opr I added the proposed changes to the branch.

**Comment by valiceops** [comment]
Just chiming in to see if there is any expected activity on this as we too are needing an updated filter to use product images not stored in the WP media library.

**Comment by opr** [comment]
Thanks for the ping @valiceops and thanks for your patience @synercoder I appreciate it has been a long time since this had activity from our end. On that note, are you still interested in working on this? If so, one thing I'd love to see added here are PHP Unit tests. We can help you with these if you're unfamiliar.

I'd recommend rebasing this on `woocommerce/trunk` too before proceeding.

There are also some code style issues that need to be solved - your code should adhere to https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/

**Comment by synercoder** [comment]
@opr I would be interested, but unfortunatly I don't have time for it any time soon, because of other commitments. Besides the time, PHP is not my main language, and I have never written a unit test in PHP, so my progress on it would be slow (again time...). If anybody could pickup the work from here that would be awesome, and if I need to do anything to make that happen, please let me know.

**Comment by opr** [comment]
@synercoder thanks for being up front and letting us know about your availability. In this case, what I suggest is: change the base of this PR to `add/cart-image-filtering` and merge it. That way you can still get credit for your contribution but we can continue to work on it in our repo (rather than on your fork). I can take care of this and I will also edit the body of your PR so it doesn't close #42259 as we will need to keep working on this.

**Comment by opr** [comment]
New PR created https://github.com/woocommerce/woocommerce/pull/52310

Thanks again @synercoder for the work you put into this so far, it is appreciated, and so is your patience while this went without comms from our end. @valiceops please feel free to follow #52310 for further updates.

**Pull request merged**